### PR TITLE
[WIP] Change cursor shape immediately on mode change

### PR DESCRIPTION
**Pull request by fmoralesc**

### CODE REVIEW

**Comment by fmoralesc** [comment]
This is causing some pathologic behavior in the travis tests, I'm not sure why. The tests all pass locally, and QuickBuild also succeeds.

**Comment by fmoralesc** [comment]
Managed to fix the issue with the tests by guarding the `ui_refresh()` call in a check for the existence of `NVIM_TUI_ENABLE_CURSOR_SHAPE`.

**Comment by justinmk** [comment]
What's this PR for? Why wasn't this needed before https://github.com/neovim/neovim/commit/ccebc1f8d3b30536bbaab532cc5acc7ba7ec3d4b ?

Forcing a refresh should be a last resort in most cases...

**Comment by fmoralesc** [comment]
This was always needed, in my experience. Using gnome-terminal at least, and xterm-256color as $TERM, the cursor didn't update when changing the mode when using `NVIM_TUI_ENABLE_CURSOR_SHAPE`.

**Comment by justinmk** [comment]
Is there a more surgical way of refreshing only the cursor, instead of the entire screen?

**Comment by fmoralesc** [comment]
Maybe calling `unibi_goto()` again to reset the cursor. I tried something similar in my tests, but the current method was the only one that worked to update the cursor. I'll try `tui_cursor_goto()`, that's something I didn't test.

**Comment by fmoralesc** [comment]
I was testing different approaches to this last night, but I found something surprising. If I change the `ui_refresh()` call with:
~~~ c
    invalidate(ui, data->row, data->row, data->col, data->col);
~~~
so it invalidates the current cell (where the cursor should be), the change of the cursor shape is immediate, but only if *not* entering insert mode, which was kind of the point. 

I also tried the `tui_eol_clear()`, which should redraw the whole line (in case the invalidated region wasn't correct, although I tried different offsets), but that seems to trigger the same behaviour.

Any ideas? /ping @tarruda

**Comment by justinmk** [comment]
> so it invalidates the current cell (where the cursor should be), the change of the cursor shape is immediate

Just guessing, but maybe try +/- 1 on the column? E.g.:

    invalidate(ui, data->row, data->row, data->col - 1, data->col - 1);

or variations thereof. Because when insert mode is entered, the cursor is "between" columns, it may be to the left of where it was in normal mode.

**Comment by fmoralesc** [comment]
Yes, I tried with `data->col -1, data->col` and `data->col, data->col +1`, but I'll try your suggestions nevertheless. I found some of these might cause glitches.

**Comment by tarruda** [comment]
Considering pangoterm(and other UIs) have no problem in updating the shape immediately, it could be a gnome-terminal bug. 

`ui_refresh` is a harsh since it causes a full redraw, but some workaround is needed. 
Maybe redrawing the current cell immediately will force gnome-terminal to update the shape:

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 56bdc73..3ef4eb7 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -425,6 +425,14 @@ static void tui_mode_change(UI *ui, int mode)
     }
   }
   data->showing_mode = mode;
+
+  // only if enabled
+  if (os_getenv("NVIM_TUI_ENABLE_CURSOR_SHAPE")) {
+    out(ui, "\b \b", 2);
+    flush_buf(ui);
+    Cell *cell = data->screen[data->row] + data->col;
+    print_cell(ui, cell);
+  }
 }

 static void tui_set_scroll_region(UI *ui, int top, int bot, int left,
```

What version are you using? My gnome terminal doesnt seem to accept the shape escape sequences so I cant test

**Comment by justinmk** [comment]
If all this does is redraw the current cell, the check for `os_getenv("NVIM_TUI_ENABLE_CURSOR_SHAPE")` can be left out.

**Comment by fmoralesc** [comment]
@tarruda Thanks for taking a look. I'm using gnome-terminal 3.16.2 currently. I'll update this tomorrow when I have some time.

**Comment by fmoralesc** [comment]
@tarruda That didn't help either, it's pretty similar to the other approaches I took invalidating the cell and the line (the cursor gets updated inmediately when entering replace mode, but not when entering insert mode) :/

**Comment by tarruda** [comment]
@fmoralesc if this is a gnome-terminal bug, then we should be able to reproduce it outside nvim. Does the cursor shape changes when you run this script?

```python
#!/usr/bin/env python
import sys
import time

time.sleep(2)
sys.stdout.write('\x1b[5 q')
sys.stdout.flush()
time.sleep(2)
sys.stdout.write('\x1b[2 q')
sys.stdout.flush()
time.sleep(2)
```

**Comment by fmoralesc** [comment]
@tarruda It does, but it changes to the bar first and the it gets hidden. So indeed, this must be a gnome-terminal bug. I tried with xterm and urxvt, and it turns out xterm doesn't change the cursor (it gets hidden) on default settings, but urxvt handles things fine.

**Comment by fmoralesc** [comment]
I opened a ticket about this at GNOME's bugzilla: https://bugzilla.gnome.org/show_bug.cgi?id=753645

**Comment by egmontkob** [comment]
This was indeed a bug in gnome-terminal, more precisely, its vte widget. The fix will appear in vte 0.40.3 and 0.42. Thanks for catching this issue!

**Comment by fmoralesc** [comment]
Updated the PR to use tarruda's variant, so it only redraws the cursor cell. Let's for @egmontkob's fix to land in VTE before we can proceed with this (it might turn out it isn't necessary).

Thanks a lot for the quick resolution on your end, @egmontkob, it's very appreciated. :+1:

**Comment by egmontkob** [comment]
You're welcome :)

If the UI refresh is meant solely as a workaround for gnome-terminal then forget it, it won't work. I can't think of any workaround (other than not changing the cursor). There's this internal boolean stating whether the blinking cursor is in "on" or "off" state at the moment, and when switching off blinking, it might get permanently stuck in "off". There's no way to directly query or modify this internal boolean. And even if you redraw everything, it'll still redraw the screen with the cursor being invisible.

The reason why we haven't discovered it is that pressing a key causes the cycle to restart from its "on" state, so if the cursor change is triggered by a keypress without a noticeable delay then it will get stuck in "on" state. I'm not sure what's the use case in nvim that causes the cursor to change with a delay from the last keypress.

**Comment by tarruda** [comment]
@egmontkob thanks. @fmoralesc I think it is best to wait  for gnome terminal to fix the issue instead of adding a workaround that doesn't solve it.

>  I'm not sure what's the use case in nvim that causes the cursor to change with a delay from the last keypress.

Nvim has communication sockets that can be used to insert keys in the input buffer, so not all output will be generated from input to the terminal.

**Comment by fmoralesc** [comment]
> @egmontkob thanks. @fmoralesc I think it is best to wait for gnome terminal to fix the issue instead of adding a workaround that doesn't solve it.

Sure, that's what I meant ;)

**Comment by liujoey** [comment]
Change the cursor shape to blinking block will partially solve this problem in gnome-terminal. Because when the cursor start blinking, it force the terminal to update the cursor.

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index c87f6d3..666042d 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -837,7 +837,7 @@ static void fix_terminfo(TUIData *data)
     data->unibi_ext.enter_replace_mode = (int)unibi_add_ext_str(ut, NULL,
         TMUX_WRAP("\x1b[3 q"));
     data->unibi_ext.exit_insert_mode = (int)unibi_add_ext_str(ut, NULL,
-        TMUX_WRAP("\x1b[2 q"));
+        TMUX_WRAP("\x1b[1 q"));
   }
 
 end:

```

**Comment by fmoralesc** [comment]
I updated this to @liujoey's proposal. It seems to work nicely (there is still a delay, though), and it is compatible with vte4 0.42 even.

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -862,7 +862,7 @@ static void fix_terminfo(TUIData *data)
     data->unibi_ext.enter_replace_mode = (int)unibi_add_ext_str(ut, NULL,
                                                                 "\x1b[3 q");
     data->unibi_ext.exit_insert_mode = (int)unibi_add_ext_str(ut, NULL,
-                                                              "\x1b[2 q");
+                                                              "\x1b[1 q");
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: line 855 checks a minimum `vte_version`. But for the change in this PR, we want to check for a maximum version.

Comment (fmoralesc) [inline src/nvim/tui/tui.c]: :+1:

Comment (fmoralesc) [inline src/nvim/tui/tui.c]: Actually, it seems like it doesn't make a difference in vte3 0.42.5; both values behave correctly.

Comment (justinmk) [inline src/nvim/tui/tui.c]: Yes, based on https://github.com/neovim/neovim/pull/3110#issuecomment-150296955 there wouldn't seem a reason for the work-around to _stop_ working at 0.42.5.  But do we want to applying  this work-around permanently?

**Comment by egmontkob** [comment]
I personally (although I'm not a neovim user :P) don't like the idea of choosing the desired look around a vte bug. The desired look should be chosen first (apparently it's a solid block and a blinking I-Beam) and then vte worked around (if possible; worst case: choose a different cursor layout only for buggy vte releases).

Above I was at the false impression that blinking restarted on keypress. This is not true, blinking is restarted if the cursor no longer blinks and if some output is produced. Based on this, I have an idea how to workaround...

Also note that once we're thinking about a workaround, I'd like to address this one as well as issue #4395. Having a common workaround for both might be simpler than a separate one for each.

Hang on...

**Comment by egmontkob** [comment]
Oh, my idea doesn't work. Nevermind. (I thought a couple of well timed back-n-forth changes to the cursor could fix it, but it can't.)

That being said, the current bug only occurs when the mode change is not the direct response to a keypress. This I believe is a quite rare event. So I think #4395 would be more important to work around than this one, if you're considering workarounds at all.

**Comment by fmoralesc** [comment]
I agree with @egmontkob, and also don't like the idea of the workaround. In my testing, this issue (the cursor not changing from insert to normal mode) is no longer present over vte3 0.42.5, so I consider it fixed (I only cared about the case where the shape didn't follow the keypresses).

Adding a backwards compatibility workaround for an issue that is present only in old versions (however recent) of the vte library feels like adding cruft to the codebase.

**Comment by fmoralesc** [comment]
Wait... doesn't the cursor initially blink? So what's the expected behavior in normal mode, a blinking or a static block?

**Comment by egmontkob** [comment]
What do you mean by "initally"? See also #4396.

As I keep toggling between the two modes, the cursor is steady block in command mode, and blinking i-beam in insert mode. (This is the actual behavior I'm seeing now; no clue what the desired behavior is.)

**Comment by egmontkob** [comment]
If I were to design this, I'd say neovim should only change the cursor shape, not the blinking. It should leave blinking on/off according to the user's global preferences. Maybe, maybe force blinking with i-beam since the i-beam cursor is hard to spot on the screen, but never force non-blinking.

Alas, the DECSCUSR escape sequences don't make it possible. See VTE enhancement request at https://bugzilla.gnome.org/show_bug.cgi?id=763935.

**Comment by fmoralesc** [comment]
@egmontkob Yes, exactly that. In my setup, when nvim starts (so in normal mode) the cursor blinks, when I enter insert it becomes a blinking i-beam and when i leave insert mode it becomes a non-blinking block. This is certainly inconsistent, and the reason I didn't initially see #3110 was also changing the cursor blinking mode.

**Comment by fmoralesc** [comment]
And actually, not only that, but after some cycles of blinking, it stops until I refocus the window. I imagine this is inherited from the terminal or vte.

**Comment by egmontkob** [comment]
It's intentional that VTE follows the GNOME-wide dconf settings "cursor-blink", "cursor-blink-time" and "cursor-blink-timeout" to be consistent with the rest of the GNOME apps. It stops blinking after a while (or on focus-out), and restarts on keypress, data received, or focus-in. (It's also buggy a little bit: https://bugzilla.gnome.org/show_bug.cgi?id=763936.) You shouldn't care about this :-)

**Comment by fmoralesc** [comment]
Sure, but as you point out, it has a bit to do with our initialization story. WIthin nvim, I think a non-blinking cursor in normal mode is best, since it makes explicit the contents of what I like to call [the "dot"](http://doc.cat-v.org/plan_9/4th_edition/papers/sam/) (the cursor/selection construct).

**Comment by justinmk** [comment]
@fmoralesc Based on the above conversation, is this PR still needed?

**Comment by fmoralesc** [comment]
I don't think so. I kept it open because of the discussion about cursor shape initialization only.

**Pull request closed without merging**